File: baekjoon/dynamic/2228/main.cpp
#include <iostream>
#include <cstdio>
#include <cstring>

using namespace std;
const int oo = 0x3f3f3f3f;

int map[100];
int cache[101][51][2];
int N,M;
int solve(int n, int m, int selected);
int main()
{
    memset(cache, -1, sizeof(cache));
    scanf("%d %d", &N, &M);
    for(int i=0; i<N; i++) {
        scanf("%d", &map[i]);
    }
    printf("%d\n", max(solve(1,0,0),solve(1,1,1)+map[0]));
    return 0;
}
           
int solve(int n, int m, int selected)
{
    if(n == N) {
        if(m == M) return 0;
        else return -oo;
    }
    if(m > M) return -oo;
    
    int& ret = cache[n][m][selected];
    if(ret != -1) return ret;
    ret = -oo;
    
    if(selected == 0) {
        ret = max(ret, solve(n+1, m+1, 1)+map[n]);
        ret = max(ret, solve(n+1, m, 0));
    } else {
        ret = max(ret, solve(n+1, m, 1)+map[n]);
        ret = max(ret, solve(n+1, m, 0));
    }
    
    return ret;
}

File: baekjoon/dynamic/2228/Main.java
import java.util.*;

class Main {
    public static final int oo = 0x3f3f3f3f;
    public static int N, M;
    public static int[] map;
    public static int[][][] cache;
    
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N = sc.nextInt();
        M = sc.nextInt();
        map = new int[N];
        cache = new int[N+1][M+1][2];
        for(int i=0; i<=N; i++) {
            for(int j=0; j<=M; j++) Arrays.fill(cache[i][j], -1);
        }
        for(int i=0; i<N; i++) {
            map[i] = sc.nextInt();
        }
        System.out.println(max(solve(1,0,0),solve(1,1,1)+map[0]));
    }
    
    public static int solve(int n, int m, int selected) {
        if(n == N) {
            if(m == M) return 0;
            else return -oo;
        }
        if(m > M) return -oo;
        
        if(cache[n][m][selected] != -1) return cache[n][m][selected];
        cache[n][m][selected] = -oo;
        
        if(selected == 0) {
            cache[n][m][selected] = max(cache[n][m][selected], solve(n+1,m,0));
            cache[n][m][selected] = max(cache[n][m][selected], solve(n+1,m+1,1)+map[n]);
        } else {
            cache[n][m][selected] = max(cache[n][m][selected], solve(n+1,m,1)+map[n]);
            cache[n][m][selected] = max(cache[n][m][selected], solve(n+1,m,0));
        }
        return cache[n][m][selected];
    }
    
    public static int max(int a, int b) {
        return a>b? a:b;
    }
}